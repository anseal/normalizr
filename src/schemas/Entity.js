import { isImmutable } from './ImmutableUtils.js'

const getDefaultGetId = (idAttribute) => (input) =>
	isImmutable(input) ? input.get(idAttribute) : input[idAttribute]

export const compileSchema = (schema) => {
	// TODO: looks like monkey-patching
	if (typeof schema === 'object' && (!schema.normalize || typeof schema.normalize !== 'function')) {
		if( Array.isArray(schema) ) {
			// TODO: schema.length === 0 ?
			if (schema.length > 1) {
				throw new Error(`Expected schema definition to be a single schema, but found ${schema.length}.`)
			}
			schema = new ArraySchema(schema[0], undefined, false)
		} else {
			schema = new ObjectSchema(schema)
		}
	}
	return schema
}

export class EntitySchema {
	constructor(key, definition = {}, options = {}) {
		if (!key || typeof key !== 'string') {
			throw new Error(`Expected a string key for Entity, but found ${key}.`)
		}

		const {
			idAttribute = 'id',
			mergeStrategy = (entityA, entityB) => ({ ...entityA, ...entityB }), // TODO: `Object.assign()` or even `retrun entityA` as a default
			processStrategy = (input) => ({ ...input }), // TODO: don't copy, at least before merge/return?
		} = options

		this._key = key
		this._getId = typeof idAttribute === 'function' ? idAttribute : getDefaultGetId(idAttribute)
		this._idAttribute = idAttribute
		this._mergeStrategy = mergeStrategy
		this._processStrategy = processStrategy
		this.define(definition)
	}

	get key() {
		return this._key
	}

	get idAttribute() {
		return this._idAttribute
	}

	define(definition) {
		// TODO: check if `definition` is an object?
		// TODO: check if it's a plain object, so that it's safe to iterate over schema without `hasOwnProperty`
		for(const key in definition) { definition[key] = compileSchema(definition[key]) }
		this.schema = Object.assign(this.schema || {}, definition);
	}

	getId(input, parent, key) {
		return this._getId(input, parent, key)
	}

	normalize(input, parent, key, entities, visited) {
		// TODO: why `!input` and not `input === null` ? because `0` will be returned and `1` will be `normalize`d
		if (typeof input !== 'object' || !input) {
			return input
		}
		const id = this.getId(input, parent, key)
		const entityType = this.key

		if(visited(input, entityType, id)) { return id }

		const processedEntity = this._processStrategy(input, parent, key)
		for(const key in this.schema) {
			if (processedEntity.hasOwnProperty(key) && typeof processedEntity[key] === 'object') { // TODO: switch places
				const schema = this.schema[key]
				const resolvedSchema = typeof schema === 'function' ? schema(input) : schema // TODO: function instead of if?
				processedEntity[key] = resolvedSchema.normalize(
					processedEntity[key],
					processedEntity,
					key,
					entities,
					visited
				)
			}
		}

		if (entityType in entities === false) {
			entities[entityType] = {}
		}
		const entitiesOfKind = entities[entityType]
	
		const existingEntity = entitiesOfKind[id]
		if (existingEntity) {
			entitiesOfKind[id] = this._mergeStrategy(existingEntity, processedEntity)
		} else {
			entitiesOfKind[id] = processedEntity
		}
		return id
	}
}

export class ObjectSchema {
	constructor(definition) {
		this.define(definition)
	}

	// TODO: DRY with EntitySchema.define?
	// TODO: now there is a difference, should we compile children schemas and should there be more tests?
	define(definition) {
		// TODO: check if `definition` is an object?
		// TODO: check if it's a plain object, so that it's safe to iterate over schema without `hasOwnProperty`
		this.schema = Object.assign(this.schema || {}, definition);
	}

	normalize(input, parent, key, entities, visited) {
		// TODO: why `!input` and not `input === null` ? because `0` will be returned and `1` will be `normalize`d
		if (typeof input !== 'object' || !input) {
			return input
		}
		// TODO: don't copy... why do we even need this? just iterate other `input`s keys and check if there's a schema for this key
		const object = { ...input }
		for(const key in this.schema) {
			const localSchema = this.schema[key] // TODO: DRY with `EntitySchema.normalize`
			const resolvedLocalSchema = typeof localSchema === 'function' ? localSchema(input) : localSchema
			const value = resolvedLocalSchema.normalize(input[key], input, key, entities, visited)
			if (value === undefined || value === null) { // TODO: options?
				delete object[key] // TODO: delete's are evil
			} else {
				object[key] = value
			}
		}
		return object
	}
}

class PolymorphicSchema {
	constructor(definition, schemaAttribute) {
		if (schemaAttribute) {
			this._schemaAttribute = typeof schemaAttribute === 'string' ? (input) => input[schemaAttribute] : schemaAttribute
			this._normalizeValue = this._normalizeValue2
		} else {
			this._normalizeValue = this._normalizeValue1
		}
		this.define(definition)
	}

	define(definition) {
		this.schema = definition
	}

	_normalizeValue1(value, parent, key, entities, visited) {
		if (!this.schema) { // TODO: _normalizeValue3?
			return value
		}
		return this.schema.normalize(value, parent, key, entities, visited)
	}

	_normalizeValue2(value, parent, key, entities, visited) {
		// TODO: just a function whould be simpler compared to function & map
		const attr = this._schemaAttribute(value, parent, key)
		const schema = this.schema[attr]

		if (!schema) { // TODO: can we precompile this?
			return value
		}
		const normalizedValue = schema.normalize(value, parent, key, entities, visited)
		return normalizedValue === undefined || normalizedValue === null
			? normalizedValue
			: {
				id: normalizedValue,
				schema: attr
			}
	}
}

export class ValuesSchema extends PolymorphicSchema {
	normalize(input, parent, key, entities, visited) {
		// TODO: why `!input` and not `input === null` ? because `0` will be returned and `1` will be `normalize`d
		if (typeof input !== 'object' || !input) {
			return input
		}
		const output = {}
		for(const key in input) { // TODO: hasOwnProperty?
			const value = input[key]
			if( value !== undefined && value !== null ) {
				output[key] = this._normalizeValue(value, input, key, entities, visited)
			}
		}
		return output
	}
}

export class ArraySchema extends PolymorphicSchema {
	constructor(definition, schemaAttribute, filterNullish = true) {
		super(definition, schemaAttribute)
		this.filterNullish = filterNullish
	}
	normalize(input, parent, key, entities, visited) {
		// TODO: why `!input` and not `input === null` ? because `0` will be returned and `1` will be `normalize`d
		if (typeof input !== 'object' || !input) {
			return input
		}
		// TODO: what is it for? maybe change API and remove?
		if( Array.isArray(input) === false ) {
			input = Object.keys(input).map((key) => input[key]) // TODO: Object.values()
		}

		// TODO: preallocate, and then cut by length?
		const normArray = []
		for(const value of input) {
			// Special case: Arrays pass *their* parent on to their children, since there
			// is not any special information that can be gathered from themselves directly
			const normValue = this._normalizeValue(value, parent, key, entities, visited)
			// TODO: what is it for, and why here and not before `_normalizeValue`?
			// TODO: filtration of falsies present in tests, but not in docs, and I have no idea why the difference
			// between [mySchema] and schema.Array(mySchema)
			if( this.filterNullish === false || (normValue !== undefined && normValue !== null) ) {
				normArray.push(normValue)
			}
		}
		return normArray
	}
}

export class UnionSchema extends PolymorphicSchema {
	constructor(definition, schemaAttribute) {
		if (!schemaAttribute) {
			throw new Error('Expected option "schemaAttribute" not found on UnionSchema.')
		}
		super(definition, schemaAttribute)
	}

	normalize(input, parent, key, entities, visited) {
		// TODO: why `!input` and not `input === null` ? because `0` will be returned and `1` will be `normalize`d
		if (typeof input !== 'object' || !input) {
			return input
		}
		return this._normalizeValue(input, parent, key, entities, visited) // TODO: something tells me that there is already a `!input` check inside
	}
}
