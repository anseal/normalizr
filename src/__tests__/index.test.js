// eslint-env jest
import { compileSchema } from '../schemas/Entity.js'
import { normalize, schema } from '../'

describe('normalize', () => {
	;[42, null, undefined, '42', () => {}].forEach((input) => {
		test(`cannot normalize input that == ${input}`, () => {
			expect(() => normalize(input, new schema.Entity('test'))).toThrow()
		})
	})

	test('cannot normalize without a schema', () => {
		expect(() => normalize({})).toThrow()
	})

	test('cannot normalize with null input', () => {
		const mySchema = new schema.Entity('tacos')
		expect(() => normalize(null, mySchema)).toThrow(/null/)
	})

	test('normalizes entities', () => {
		const mySchema = new schema.Entity('tacos')

		expect(
			normalize(
				[
					{ id: 1, type: 'foo' },
					{ id: 2, type: 'bar' },
				],
				[mySchema]
			)
		).toMatchSnapshot()
	})

	test('normalizes entities with circular references', () => {
		const user = new schema.Entity('users')
		user.define({
			friends: [user],
		})

		const input = { id: 123, friends: [] }
		input.friends.push(input)

		expect(normalize(input, user, true)).toMatchSnapshot()
	})

	test('normalizes nested entities', () => {
		const user = new schema.Entity('users')
		const comment = new schema.Entity('comments', {
			user: user,
		})
		const article = new schema.Entity('articles', {
			author: user,
			comments: [comment],
		})

		const input = {
			id: '123',
			title: 'A Great Article',
			author: {
				id: '8472',
				name: 'Paul',
			},
			body: 'This article is great.',
			comments: [
				{
					id: 'comment-123-4738',
					comment: 'I like it!',
					user: {
						id: '10293',
						name: 'Jane',
					},
				},
			],
		}
		expect(normalize(input, article)).toMatchSnapshot()
	})

	test('does not modify the original input', () => {
		const user = new schema.Entity('users')
		const article = new schema.Entity('articles', { author: user })
		const input = Object.freeze({
			id: '123',
			title: 'A Great Article',
			author: Object.freeze({
				id: '8472',
				name: 'Paul',
			}),
		})
		expect(() => normalize(input, article)).not.toThrow()
	})

	test('ignores null values', () => {
		const myEntity = new schema.Entity('myentities')
		expect(normalize([null], [myEntity])).toMatchSnapshot()
		expect(normalize([undefined], [myEntity])).toMatchSnapshot()
		expect(normalize([false], [myEntity])).toMatchSnapshot()
	})

	// TODO: remove? basically it tests that a user can break encapsulation. why whould we allow this?!
	test('can use fully custom entity classes', () => {
		class MyEntity extends schema.Entity {
			schema = {
				children: compileSchema([new schema.Entity('children')]),
			}

			getId(entity, parent, key) {
				return entity.uuid
			}

			normalize(input, parent, key, entities, visitedEntities) {
				const entity = { ...input }
				Object.keys(this.schema).forEach((key) => {
					const schema = this.schema[key]
					entity[key] = schema.normalize(input[key], input, key, entities, visitedEntities)
				})
				// TODO: inherit this part from `EntitySchema.normalize`?
				const entityType = this.key
				const id = this.getId(entity)
				if (entityType in entities === false) {
					entities[entityType] = {}
				}
				const entitiesOfKind = entities[entityType]
			
				const existingEntity = entitiesOfKind[id]
				if (existingEntity) {
					entitiesOfKind[id] = this._mergeStrategy(existingEntity, entity)
				} else {
					entitiesOfKind[id] = entity
				}

				return {
					uuid: id,
					schema: entityType,
				}
			}
		}

		const mySchema = new MyEntity('food')
		expect(
			normalize(
				{
					uuid: '1234',
					name: 'tacos',
					children: [{ id: 4, name: 'lettuce' }],
				},
				mySchema
			)
		).toMatchSnapshot()
	})

	test('uses the non-normalized input when getting the ID for an entity', () => {
		const userEntity = new schema.Entity('users')
		const idAttributeFn = jest.fn((nonNormalized, parent, key) => nonNormalized.user.id)
		const recommendation = new schema.Entity(
			'recommendations',
			{ user: userEntity },
			{
				idAttribute: idAttributeFn,
			}
		)
		expect(normalize({ user: { id: '456' } }, recommendation)).toMatchSnapshot()
		expect(idAttributeFn.mock.calls).toMatchSnapshot()
		expect(recommendation.idAttribute).toBe(idAttributeFn)
	})

	test('passes over pre-normalized values', () => {
		const userEntity = new schema.Entity('users')
		const articleEntity = new schema.Entity('articles', { author: userEntity })

		expect(normalize({ id: '123', title: 'normalizr is great!', author: 1 }, articleEntity)).toMatchSnapshot()
	})

	test('can normalize object without proper object prototype inheritance', () => {
		const test = { id: 1, elements: [] }
		test.elements.push(
			Object.assign(Object.create(null), {
				id: 18,
				name: 'test',
			})
		)

		const testEntity = new schema.Entity('test', {
			elements: [new schema.Entity('elements')],
		})

		expect(() => normalize(test, testEntity)).not.toThrow()
	})

	test('can normalize entity nested inside entity using property from parent', () => {
		const linkablesSchema = new schema.Entity('linkables')
		const mediaSchema = new schema.Entity('media')
		const listsSchema = new schema.Entity('lists')

		const schemaMap = {
			media: mediaSchema,
			lists: listsSchema,
		}

		linkablesSchema.define({
			data: (parent) => schemaMap[parent.schema_type],
		})

		const input = {
			id: 1,
			module_type: 'article',
			schema_type: 'media',
			data: {
				id: 2,
				url: 'catimage.jpg',
			},
		}

		expect(normalize(input, linkablesSchema)).toMatchSnapshot()
	})

	test('can normalize entity nested inside object using property from parent', () => {
		const mediaSchema = new schema.Entity('media')
		const listsSchema = new schema.Entity('lists')

		const schemaMap = {
			media: mediaSchema,
			lists: listsSchema,
		}

		const linkablesSchema = {
			data: (parent) => schemaMap[parent.schema_type],
		}

		const input = {
			id: 1,
			module_type: 'article',
			schema_type: 'media',
			data: {
				id: 2,
				url: 'catimage.jpg',
			},
		}

		expect(normalize(input, linkablesSchema)).toMatchSnapshot()
	})
})
