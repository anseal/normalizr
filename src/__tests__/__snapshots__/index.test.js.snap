// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalize ignores null values 1`] = `
Object {
  "entities": Object {},
  "result": Array [
    null,
  ],
}
`;

exports[`normalize ignores null values 2`] = `
Object {
  "entities": Object {},
  "result": Array [
    undefined,
  ],
}
`;

exports[`normalize ignores null values 3`] = `
Object {
  "entities": Object {},
  "result": Array [
    false,
  ],
}
`;

exports[`normalize normalizes entities 1`] = `
Object {
  "entities": Object {
    "tacos": Object {
      "1": Object {
        "id": 1,
        "type": "foo",
      },
      "2": Object {
        "id": 2,
        "type": "bar",
      },
    },
  },
  "result": Array [
    1,
    2,
  ],
}
`;

exports[`normalize normalizes entities with circular references 1`] = `
Object {
  "entities": Object {
    "users": Object {
      "123": Object {
        "friends": Array [
          123,
        ],
        "id": 123,
      },
    },
  },
  "result": 123,
}
`;

exports[`normalize normalizes nested entities 1`] = `
Object {
  "entities": Object {
    "articles": Object {
      "123": Object {
        "author": "8472",
        "body": "This article is great.",
        "comments": Array [
          "comment-123-4738",
        ],
        "id": "123",
        "title": "A Great Article",
      },
    },
    "comments": Object {
      "comment-123-4738": Object {
        "comment": "I like it!",
        "id": "comment-123-4738",
        "user": "10293",
      },
    },
    "users": Object {
      "10293": Object {
        "id": "10293",
        "name": "Jane",
      },
      "8472": Object {
        "id": "8472",
        "name": "Paul",
      },
    },
  },
  "result": "123",
}
`;

exports[`normalize passes over pre-normalized values 1`] = `
Object {
  "entities": Object {
    "articles": Object {
      "123": Object {
        "author": 1,
        "id": "123",
        "title": "normalizr is great!",
      },
    },
  },
  "result": "123",
}
`;

exports[`normalize uses the non-normalized input when getting the ID for an entity 1`] = `
Object {
  "entities": Object {
    "recommendations": Object {
      "456": Object {
        "user": "456",
      },
    },
    "users": Object {
      "456": Object {
        "id": "456",
      },
    },
  },
  "result": "456",
}
`;

exports[`normalize uses the non-normalized input when getting the ID for an entity 2`] = `
Array [
  "[{\\"user\\":{\\"id\\":\\"456\\"}},{\\"user\\":{\\"id\\":\\"456\\"}},null]",
]
`;
